### PR TITLE
Add support for failure-mode label

### PR DESCRIPTION
**Pull request by croomes**
Adds the ability to set a volume's failure mode behaviour by updating the PVC's `storageos.com/failure-mode` label.

Also updates the go-api dependency to `v2.4.0`.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -13,7 +13,7 @@ require (
 	github.com/patrickmn/go-cache v2.1.0+incompatible
 	github.com/pkg/errors v0.9.1
 	github.com/prometheus/client_golang v1.7.1
-	github.com/storageos/go-api/v2 v2.3.1-0.20210211170707-7520f2fe0974
+	github.com/storageos/go-api/v2 v2.3.1-0.20210420150320-a30cf41359d2
```
Comment (croomes) [inline go.mod:16]: I had trouble setting to `v2.4.0` so used the commit ref for the tag instead.  I might have something cached as I'd imported from a branch named `v2.4.0` before I tagged.

File: internal/pkg/storageos/volume_labels.go [hunk 2]
```diff
@@ -162,3 +180,77 @@ func (c *Client) EnsureReplicas(ctx context.Context, key client.ObjectKey, desir
...
+
+// EnsureFailureMode ensures that the desired failure mode has been applied to
+// the StorageOS volume.
+func (c *Client) EnsureFailureMode(ctx context.Context, key client.ObjectKey, desired string) error {
+	funcName := "ensure_failure_mode"
+	start := time.Now()
+	defer func() {
+		metrics.Latency.Observe(funcName, time.Since(start))
+	}()
+	observeErr := func(e error) error {
```
Comment (mhmxs) [inline internal/pkg/storageos/volume_labels.go:192]: Just an idea. What if we give some name to error in the signature and create a `defer` which checks returned error. If not nil then `increment`. With this helper the rest of the code should able to `return err` like any other regular code, while we have the incrementation too.

File: internal/pkg/storageos/volume_labels.go [hunk 3]
```diff
@@ -162,3 +180,77 @@ func (c *Client) EnsureReplicas(ctx context.Context, key client.ObjectKey, desir
...
+
+	vol, err := c.getVolume(ctx, key)
+	if err != nil {
+		return observeErr(err)
+	}
+
+	var current string
+
+	// If label already set, get value.
+	for k, v := range vol.Labels {
```
Comment (mhmxs) [inline internal/pkg/storageos/volume_labels.go:207]: I guess `current := vol.Labels[ReservedLabelFailureMode]` does the same.
`var current string` is an empty string while `vol.Labels[ReservedLabelFailureMode]` returns value or empty string.

File: internal/pkg/storageos/volume_labels.go [hunk 4]
```diff
@@ -162,3 +180,77 @@ func (c *Client) EnsureReplicas(ctx context.Context, key client.ObjectKey, desir
...
+		return observeErr(api.MapAPIError(err, resp))
+	}
+	return observeErr(nil)
+}
+
+// ParseFailureMode parses a string and returns either a failure mode intent or
+// a threshold, if set.  Only one of the intent or threshold should be set.
+func ParseFailureMode(s string) (api.FailureModeIntent, uint64, error) {
+	//  Return defaults if no value set.
+	if s == "" {
```
Comment (mhmxs) [inline internal/pkg/storageos/volume_labels.go:236]: This expression should move int `switch` `case "", FailureModeHard`

File: internal/pkg/storageos/volume_labels.go (hunk 2 above)
Comment (croomes) [reply internal/pkg/storageos/volume_labels.go:192]: That's a good idea.  We use the current pattern in many places, so let's do a follow-up PR after `v2.4.0` to change everywhere.

File: internal/pkg/storageos/volume_labels.go (hunk 3 above)
Comment (croomes) [reply internal/pkg/storageos/volume_labels.go:207]: yes...  I do this often - I don't know why!

File: internal/pkg/storageos/volume_labels.go (hunk 4 above)
Comment (croomes) [reply internal/pkg/storageos/volume_labels.go]: yes, that's better.

**Review by mhmxs — APPROVED** [review]

**Pull request merged**